extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -128,235 +123,268 @@ export class Bind {
   }
 
   /**
-   * Scans children for attributes that conform to bind syntax and returns
-   * all bindings.
+   * Scans `body` for attributes that conform to bind syntax and returns
+   * a tuple containing bound elements and binding data for the evaluator.
    * @param {!Element} body
-   * @return {{bindings: !Array<BindingDef>, expressions: !Array<string>}}
+   * @return {{
+   *   boundElements: !Array<BoundElementDef>,
+   *   evaluatees: !Array<./bind-evaluator.EvaluateeDef>,
+   * }}
    * @private
    */
   scanForBindings_(body) {
     // TODO(choumx): Chunk if taking too long in a single frame.
 
-    const bindings = [];
-    const expressions = [];
+    /** @type {!Array<BoundElementDef>} */
+    const boundElements = [];
+    /** @type {!Array<./bind-evaluator.EvaluateeDef>} */
+    const evaluatees = [];
+
     const elements = body.getElementsByTagName('*');
```

Comment:
I'm actually planning on taking a closer look at performance of this (per the TODO above) and other initialization tasks like the parsing of expressions. Justin also suggested using attribute existence query selector. I'll check out Treewalker too, thanks for the tip!